recursion/src/air/alu/air.rs

```diff
@@ -0,0 +1,464 @@
+
+        let (prefix, rows, suffix) = unsafe { trace.values.align_to_mut::<AluCols<F, R>>() };
+        assert!(prefix.is_empty(), "Alignment should match");
+        assert!(suffix.is_empty(), "Alignment should match");
+        assert_eq!(rows.len(), events_len);
+
+        for event in events {
+            println!("event = {:?}", *event);
+            for i in 0..R {
+                let row = &mut rows[i];
+                row.left_addr[i] = F::from_usize(event.left_addr[i]);
+                row.left_val[i] = event.left_val[i];
+                row.right_addr[i] = F::from_usize(event.right_addr[i]);
+                row.right_val[i] = event.right_val[i];
+                row.res_addr[i] = F::from_usize(event.res_addr[i]);
+                row.res_val[i] = event.res_val[i];
+            }
+        }
+        trace
+    }
+}
+
+/*
+Asserts a op b = c, where op in {+, -, *}.
+(so that the total constraint degree is self.degree).
+REPETITIONS counts how many `a * b = c` operations to do per row in the AIR
+*/
+pub trait AluAir<AB: AirBuilder, const R: usize = 1>:
+    Operation + Air<AB> + GenerateAluTrace<R>
+{
+    fn random_valid_trace<F: Field>(&self, rows: usize, valid: bool) -> RowMajorMatrix<F>
+    where
+        StandardUniform: Distribution<F>,
+    {
+        let mut rng = SmallRng::seed_from_u64(1);
+        let n_padded = rows.next_power_of_two();
+        let mut trace =
+            RowMajorMatrix::new(F::zero_vec(n_padded * Self::TRACE_WIDTH), Self::TRACE_WIDTH);
+
+        let (prefix, rows, suffix) = unsafe { trace.values.align_to_mut::<AluCols<F, R>>() };
+        assert!(prefix.is_empty(), "Alignment should match");
+        assert!(suffix.is_empty(), "Alignment should match");
+        assert_eq!(rows.len(), n_padded);
+
+        for (i, row) in rows.iter_mut().enumerate() {
+            for (left_addr, left_val, right_addr, right_val, res_addr, res_val) in izip!(
+                row.left_addr.iter_mut(),
+                row.left_val.iter_mut(),
+                row.right_addr.iter_mut(),
+                row.right_val.iter_mut(),
+                row.res_addr.iter_mut(),
+                row.res_val.iter_mut(),
+            ) {
+                *left_addr = F::from_usize(3 * i);
+                *left_val = rng.random();
+                *right_addr = F::from_usize(3 * i + 1);
+                *right_val = rng.random();
+                *res_addr = F::from_usize(3 * i + 2);
+                *res_val = self.apply(*left_val, *right_val);
+                if !valid {
+                    // make it invalid
+                    *res_val *= F::TWO;
+                }
+            }
+        }
+        trace
+    }
+
+    fn width(&self) -> usize {
+        Self::TRACE_WIDTH
+    }
+
+    fn eval(&self, builder: &mut AB) {
+        let main = builder.main();
+        let local = main.row_slice(0).expect("Matrix is empty?");
+        let local: &AluCols<AB::Var> = (*local).borrow();
```

Comment:
I don't think this is correct? this currently assumes `R = 1`, or am I missing something? If that's the case, I'd probably specify it still, out of clarity
```suggestion
        let local: &AluCols<AB::Var, R> = (*local).borrow();
```
with `R` defined somewhere, or directly `1`.